server/src/main/java/org/opensearch/tasks/TaskManager.java

```diff
@@ -448,6 +461,18 @@ public void waitForTaskCompletion(Task task, long untilInNanos) {
         throw new OpenSearchTimeoutException("Timed out waiting for completion of [{}]", task);
     }
 
+    /**
+     * Takes actions when a task is registered and its execution starts
+     *
+     * @param task getting executed.
+     * @return AutoCloseable to free up resources (clean up thread context) when task execution block returns
+     */
+    public ThreadContext.StoredContext taskExecutionStarted(Task task) {
+        if (taskResourceTrackingService.get() == null) return () -> {};
```

Comment:
Shouldn't we return the current `StoredContext` here? It looks like we have will have a brain split otherwise: `taskExecutionStarted` returns empty `StoredContext` however in fact it is not empty.

Reply:
This is sort of a freeing resources method. That this method allocated something. Once execution is finished use this reference to de-allocate.

So, here if we are not creating a new thread context so no point in giving a reference to pop it because we didn't modify anything. The original ThreadContext will get popped from where it was originally created (e.g InboundPipeline.)